chrome/content/zotero/preferences/preferences_advanced.js

```diff
@@ -579,30 +673,80 @@ Zotero_Preferences.Attachment_Base_Directory = {
+		libraries.forEach(function (library) {
+			var libraryName = library.name;
+			var libraryID = parseInt(library.libraryID);
+			var checked = Zotero.Attachments.getSaveRelativePathByLibrary(libraryID);
+			var attachmentBasePath = Zotero.Attachments.getBasePathByLibrary(libraryID);
+
+			var treeitem = document.createElement('treeitem');
+			var treerow = document.createElement('treerow');
+			var checkboxCell = document.createElement('treecell');
+			var nameCell = document.createElement('treecell');
+			var pathCell = document.createElement('treecell');
+
+			checkboxCell.setAttribute('value', checked);
+			checkboxCell.setAttribute('editable', true);
+			nameCell.setAttribute('label', libraryName);
+			nameCell.setAttribute('value', libraryID);
+			nameCell.setAttribute('editable', false);
+			if (attachmentBasePath) {
+				pathCell.setAttribute('label', attachmentBasePath);
+				pathCell.setAttribute('value', attachmentBasePath);
+			}
+			pathCell.setAttribute('editable', false);
+
+			treerow.appendChild(checkboxCell);
+			treerow.appendChild(nameCell);
+			treerow.appendChild(pathCell);
+			treeitem.appendChild(treerow);
+			treechildren.appendChild(treeitem);
+		});
+
+		// Prune preferences of any libraries that no longer exist
+		var existentLibraryIDs = libraries.map(l => parseInt(l.libraryID));
+		var savedLibraryIDs =
+				Object.keys(
+					JSON.parse(
+						Zotero.Prefs.get("librarySaveRelativeAttachmentPaths") || "{}"
+					)
+				).map(id => parseInt(id));
+		savedLibraryIDs.forEach(function (libraryID) {
+			if (existentLibraryIDs.indexOf(libraryID) == -1) {
+				Zotero.debug(`Pruning attachment base path preferences for non-existent library '${libraryID}'`);
+				Zotero.Attachments.setSaveRelativePathByLibrary(libraryID, null);
+				Zotero.Attachments.setBasePathByLibrary(libraryID, null);
+			}
+		});
 	}),
+
 	
-	
-	changePath: Zotero.Promise.coroutine(function* (basePath) {
-		Zotero.debug(`New base directory is ${basePath}`);
+	changePath: Zotero.Promise.coroutine(function* (libraryID, basePath) {
+		Zotero.debug(`Setting new base attachment directory for '${libraryID}': '${basePath}'`);
 		
-		// Find all current attachments with relative attachment paths
-		var sql = "SELECT itemID FROM itemAttachments WHERE linkMode=? AND path LIKE ?";
-		var params = [
-			Zotero.Attachments.LINK_MODE_LINKED_FILE,
-			Zotero.Attachments.BASE_PATH_PLACEHOLDER + "%"
-		];
-		var oldRelativeAttachmentIDs = yield Zotero.DB.columnQueryAsync(sql, params);
-		
-		//Find all attachments on the new base path
+		// Find all attachments on the new base path
 		var sql = "SELECT itemID FROM itemAttachments WHERE linkMode=?";
 		var params = [Zotero.Attachments.LINK_MODE_LINKED_FILE];
-		var allAttachments = yield Zotero.DB.columnQueryAsync(sql, params);
+		var allAttachmentIDs = yield Zotero.DB.columnQueryAsync(sql, params);
+		var isLibraryBasePathSet = Zotero.Attachments.getBasePathByLibrary(libraryID);
```

Comment:
Why was `getPath` (see `Zotero.Attachments.getBasePathByLibrary`) called every loop when it shouldn't change from one iteration to the next?
It's fair to check that a base attachment path is set for the library, but it doesn't need to be checked every time...